api/base/parsers.py

```diff
@@ -15,30 +18,65 @@ class JSONAPIParser(JSONParser):
     media_type = 'application/vnd.api+json'
     renderer_class = JSONAPIRenderer
 
+    @staticmethod
+    def flatten_data(resource_object, request_method, is_list=False):
+        """
+        Flattens data objects, making attributes fields the same level as id and type.
+        """
+
+        if "attributes" not in resource_object and request_method != 'DELETE':
+            raise JSONAPIException(source={'pointer': '/data/attributes'}, detail=NO_ATTRIBUTES_ERROR)
+
+        object_id = resource_object.get('id')
+        object_type = resource_object.get('type')
+
+        # For validating type and id for bulk delete:
+        if is_list and request_method == 'DELETE':
+            if object_id is None:
+                raise JSONAPIException(source={'pointer': '/data/id'}, detail=NO_ID_ERROR)
+
+            if object_type is None:
+                raise JSONAPIException(source={'pointer': '/data/type'}, detail=NO_TYPE_ERROR)
+
+        attributes = resource_object.get('attributes')
+        parsed = {'id': object_id, 'type': object_type}
+        if attributes:
+            parsed.update(attributes)
+
+        return parsed
+
+    # Overrides JSONParser
     def parse(self, stream, media_type=None, parser_context=None):
         """
-        Parses the incoming bytestream as JSON and returns the resulting data
+        Parses the incoming bytestream as JSON and returns the resulting data.
         """
         result = super(JSONAPIParser, self).parse(stream, media_type=media_type, parser_context=parser_context)
         if not isinstance(result, dict):
             raise ParseError()
         data = result.get('data', {})
 
+        method = parser_context['request'].method
         if data:
-            if 'attributes' not in data:
-                raise JSONAPIException(source={'pointer': '/data/attributes'}, detail=NO_ATTRIBUTES_ERROR)
-            id = data.get('id')
-            object_type = data.get('type')
-            attributes = data.get('attributes')
+            if is_bulk_request(stream):
```

Comment:
Where else can I access the request here? It does work, with  stream.META.get('CONTENT_TYPE')